crates/node/builder/src/launch/common.rs

```diff
@@ -843,6 +869,102 @@ where
     consensus: Arc<dyn Consensus>,
 }
 
+impl<DB, T, CB> InitializedComponents for WithComponents<DB, T, CB>
+where
+    T: FullNodeTypes<Provider = BlockchainProvider<DB>>,
+    CB: NodeComponentsBuilder<T>,
+{
+    type Node = NodeAdapter<T, CB::Components>;
+
+    fn head(&self) -> Head {
+        self.head
+    }
+
+    fn blockchain_db(&self) -> &BlockchainProvider<Self::Node::DB> {
+        &self.blockchain_db
+    }
+
+    fn consensus(&self) -> Arc<dyn Consensus> {
+        self.consensus.clone()
+    }
+
+    fn sync_metrics_tx(&self) -> UnboundedSender<MetricEvent> {
+        self.db_provider_container.metrics_sender.clone()
+    }
+
+    fn tree_config(&self) -> &BlockchainTreeConfig {
+        &self.tree_config
+    }
+
+    fn components(&self) -> &Self::Node {
+        &self.node_adapter
+    }
+}
+
+#[allow(missing_debug_implementations)]
+#[derive(Deref)]
+pub struct WithComponentsExt<Rpc> {
+    #[deref]
+    core: Box<dyn InitializedComponents>,
+    rpc: Option<Rpc>,
+    // pipeline: Pipeline,
+    // beacon_engine: Engine,
+}
+
+pub trait ExtComponentBuilder {
+    type Node: FullNodeComponentsExt;
+
+    fn build(self) -> Self::Node;
+}
+
+impl<Rpc> ExtComponentBuilder for WithComponentsExt<Rpc>
+where
+    Rpc: Rpc<Self::Node>,
+{
+    type Node = NodeAdapterExt<<Self as InitializedComponents>::Node, Rpc>;
+
+    fn build(self) -> Self::Node {
+        let Self { core, rpc } = self;
+
+        NodeAdapterExt { core, rpc }
+    }
+}
+
+pub trait InstallPipeline: ExtComponentBuilder {}
+pub trait InstallBeaconEngine: ExtComponentBuilder + InstallPipeline {}
+pub trait InstallRpc: ExtComponentBuilder + InstallBeaconEngine {
+    // fn beacon_engine_handle(&self) -> &Self::Node::Engine::Handle;
+    fn beacon_engine_handle(&self) -> &BeaconConsensusEngineHandle<Self::Node::Engine>;
```

Comment:
will just move this getter into `RpcBuilder` trait, and same for other dependencies getters, and delete these `Install` traits